sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubClient.java

```diff
@@ -0,0 +1,176 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.dataflow.sdk.io;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * A helper interface for talking to pub/sub via an underlying transport.
+ */
+public interface PubsubClient extends AutoCloseable {
+  /**
+   * Gracefully close the underlying transport.
+   */
+  @Override
+  void close();
+
+  /**
+   * A message to be sent to pub/sub.
+   */
+  class OutgoingMessage {
+    /**
+     * Underlying (encoded) element.
+     */
+    public final byte[] elementBytes;
+
+    /**
+     * Timestamp for element (ms since epoch).
+     */
+    public final long timestampMsSinceEpoch;
+
+    public OutgoingMessage(byte[] elementBytes, long timestampMsSinceEpoch) {
+      this.elementBytes = elementBytes;
+      this.timestampMsSinceEpoch = timestampMsSinceEpoch;
+    }
+  }
+
+  /**
+   * A message received from pub/sub.
+   */
+  class IncomingMessage {
+    /**
+     * Underlying (encoded) element.
+     */
+    public final byte[] elementBytes;
+
+    /**
+     * Timestamp for element (ms since epoch). Either pub/sub's processing time,
+     * or the custom timestamp associated with the message.
+     */
+    public final long timestampMsSinceEpoch;
+
+    /**
+     * Timestamp (in system time) at which we requested the message (ms since epoch).
+     */
+    public final long requestTimeMsSinceEpoch;
+
+    /**
+     * Id to pass back to pub/sub to acknowledge receipt of this message.
+     */
+    public final String ackId;
+
+    /**
+     * Id to pass to the runner to distinguish this message from all others.
+     */
+    public final byte[] recordId;
+
+    public IncomingMessage(
+        byte[] elementBytes,
+        long timestampMsSinceEpoch,
+        long requestTimeMsSinceEpoch,
+        String ackId,
+        byte[] recordId) {
+      this.elementBytes = elementBytes;
+      this.timestampMsSinceEpoch = timestampMsSinceEpoch;
+      this.requestTimeMsSinceEpoch = requestTimeMsSinceEpoch;
+      this.ackId = ackId;
+      this.recordId = recordId;
+    }
+  }
+
+  /**
+   * Publish {@code outgoingMessages} to pub/sub {@code topic}. Return number of messages
+   * published.
+   *
+   * @throws IOException
+   */
+  int publish(String topic, Iterable<OutgoingMessage> outgoingMessages) throws IOException;
+
+  /**
+   * Request the next batch of up to {@code batchSize} messages from {@code subscription}.
+   * Return the received messages, or empty collection if none were available. Does not
+   * wait for messages to arrive. Returned messages will record heir request time
+   * as {@code requestTimeMsSinceEpoch}.
+   *
+   * @throws IOException
+   */
+  Collection<IncomingMessage> pull(
```

Comment:
It would seem pretty normal to have explicit `Subscription` class rather than passing the string name as a key each time. Is there a reason this cannot be done?

For example, if `#pull` did not take a subscription string, but pulled all available messages from all subscriptions in bulk, then I would understand.

Related - publishing and subscribing seem to be two separable activities for pub/sub but this client seems like roughly the (disjoint) union of the two.